File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
using namespace std;

int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};

int solution(vector<vector<int>> maps) 
{
    int n = maps.size();
    int m = maps[0].size();
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    queue<pair<int, int>> q;

    q.push({0, 0});
    visited[0][0] = true;

    while (!q.empty()) 
    {
        int y = q.front().first;
        int x = q.front().second;
        q.pop();

        for (int i = 0; i < 4; i++) 
        {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny >= 0 && ny < n && nx >= 0 && nx < m) 
            {
                if (!visited[ny][nx] && maps[ny][nx] == 1) 
                {
                    visited[ny][nx] = true;
                    maps[ny][nx] = maps[y][x] + 1;  
                    q.push({ny, nx});
                }
            }
        }
    }

    int result = maps[n-1][m-1];
    return (result == 1) ? -1 : result;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;

class Solution
{
    public int solution(int[,] maps) 
    {
        int n = maps.GetLength(0); // 행
        int m = maps.GetLength(1); // 열

        int[] dx = { 1, -1, 0, 0 };
        int[] dy = { 0, 0, 1, -1 };

        bool[,] visited = new bool[n, m];
        Queue<(int y, int x, int cost)> q = new Queue<(int, int, int)>();

        q.Enqueue((0, 0, 1));
        visited[0, 0] = true;

        while (q.Count > 0)
        {
            var (curY, curX, curCost) = q.Dequeue();

            if (curY == n - 1 && curX == m - 1) return curCost;

            for (int i = 0; i < 4; i++)
            {
                int ny = curY + dy[i];
                int nx = curX + dx[i];

                if (ny >= 0 && nx >= 0 && ny < n && nx < m)
                {
                    if (!visited[ny, nx] && maps[ny, nx] == 1)
                    {
                        visited[ny, nx] = true;
                        q.Enqueue((ny, nx, curCost + 1));
                    }
                }
            }
        }

        return -1; // 도착 불가
    }
}
